File: 2021/Day 17/TS/part1.ts
import fs from 'fs'
import path from 'path'

const file: string = fs.readFileSync(path.resolve(__dirname, '../data.txt'), 'utf-8')
const bounds = file.split(',').map((x) =>
    x
        .split('=')[1]
        .split('..')
        .map((n) => parseInt(n)),
)
const target = { x: { max: bounds[0][1], min: bounds[0][0] }, y: { max: bounds[1][1], min: bounds[1][0] } }

let start = performance.now()
let i = 0
let height
while (true) {
    height = 0.5 * i * (i - 1)
    let mostStepsPossible = 0.5 * (1 + Math.sqrt(8 * height + -8 * target.y.min + 1))
    if (mostStepsPossible == i + 1) break
    i++
}

console.log('Highest possible shot:', height)
console.log('Solved in:', performance.now() - start + 'ms')

File: 2021/Day 17/C++/part1.cpp
#include <math.h>

#include <chrono>
#include <fstream>
#include <iostream>
#include <sstream>

// Namespaces
using namespace std;  // Bad practice, I know

struct Target {
    pair<int, int> x;
    pair<int, int> y;
};

// Main
int main() {
    pair<int, int> x;
    pair<int, int> y;

    // Get input
    ifstream file("../data.txt");
    string line;
    getline(file, line);
    stringstream ss{line};
    ss.ignore(99, '=') >> x.first;
    ss.ignore(2) >> x.second;
    ss.ignore(99, '=') >> y.first;
    ss.ignore(2) >> y.second;
    Target target{x, y};

    // Find highest point
    auto start = chrono::high_resolution_clock::now();
    int i = 0;
    int height;
    while (true) {
        height = 0.5 * i * (i - 1);
        double mostSteps = 0.5 * (1 + sqrt(8 * height - 8 * target.y.first + 1));
        if (mostSteps == i + 1) break;
        i++;
    }
    auto stop = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::microseconds>(stop - start);

    cout << "Highest possible shot: " << height << endl;
    cout << "Time elapsed: " << duration.count() << "us" << endl;

    return 0;
}